math/vc_aos/include/algebra/math/impl/vc_aos_transform3.hpp

```diff
@@ -184,88 +210,107 @@ struct transform3 {
   ///
   /// @param m is the rotation matrix
   /// @param v is the vector to be rotated
-  ALGEBRA_HOST_DEVICE
-  inline constexpr auto rotate(const matrix44 &m, const vector3 &v) const {
+  template <typename vector3_type>
+  ALGEBRA_HOST_DEVICE inline constexpr auto rotate(
+      const matrix44 &m, const vector3_type &v) const {
 
     return m.x * v[0] + m.y * v[1] + m.z * v[2];
   }
 
-  /// @returns the translation of the transform
+  /// This method retrieves the rotation of a transform
+  ALGEBRA_HOST_DEVICE
+  inline array_type<16> rotation() const {
+
+    array_type<16> submatrix;
+    for (unsigned int irow = 0; irow < 3; ++irow) {
+      for (unsigned int icol = 0; icol < 3; ++icol) {
+        submatrix[icol + irow * 4] = element_getter()(_data, irow, icol);
+      }
+    }
+    return submatrix;
+  }
+
+  /// This method retrieves x axis
   ALGEBRA_HOST_DEVICE
-  inline point3 translation() const { return _data.t; }
+  inline const auto &x() const { return _data.x; }
 
-  /// @returns the 4x4 matrix of the transform
+  /// This method retrieves y axis
+  ALGEBRA_HOST_DEVICE
+  inline const auto &y() const { return _data.y; }
+
+  /// This method retrieves z axis
+  ALGEBRA_HOST_DEVICE
+  inline const auto &z() const { return _data.z; }
+
+  /// This method retrieves the translation
+  ALGEBRA_HOST_DEVICE
+  inline const auto &translation() const { return _data.t; }
+
+  /// This method retrieves the 4x4 matrix of a transform
   ALGEBRA_HOST_DEVICE
   inline const matrix44 &matrix() const { return _data; }
 
-  /// @returns the 4x4 matrix of the inverse transform
+  /// This method retrieves the 4x4 matrix of an inverse transform
   ALGEBRA_HOST_DEVICE
   inline const matrix44 &matrix_inverse() const { return _data_inv; }
 
-  /// This method transforms a point from the local 3D cartesian frame
-  /// to the global 3D cartesian frame
+  /// This method transform from a point from the local 3D cartesian frame
+  ///  to the global 3D cartesian frame
   ///
-  /// @tparam point3_t 3D point
+  /// @tparam point_type 3D point
   ///
   /// @param v is the point to be transformed
   ///
   /// @return a global point
-  template <
-      typename point3_t,
-      std::enable_if_t<std::is_convertible_v<point3_t, point3>, bool> = true>
-  ALGEBRA_HOST_DEVICE inline point3 point_to_global(const point3_t &v) const {
+  template <typename point3_type>
+  ALGEBRA_HOST_DEVICE inline auto point_to_global(const point3_type &p) const {
 
-    return rotate(_data, v) + _data.t;
+    return _data.x * p[0] + _data.y * p[1] + _data.z * p[2] + _data.t;
```

Comment:
It's a pity that we couldn't keep function calls like `cross(...)` (from an earlier part in the code) and `rotate(...)` here alive. 😦 I imagine that there's a technical reason for this, but it does hurt readability.

Reply:
Yes, for ```cross``` there was a reason (which I forgot), but I remember being unhappy with that change, as well. As for the rotation part, not sure why I changed that...

Reply:
The cross function is plugin specific, so if I include the vector function implementations here, the two plugins would become dependent on the presence of the other. But I put rotate function back in